### PR TITLE
Settle unlock tests refactor

### PR DESCRIPTION
**Pull request by loredanacirstea**
fixes https://github.com/raiden-network/raiden-contracts/issues/132

Linked issue: https://github.com/raiden-network/raiden-contracts/issues/188. This PR implements tests for old balance proofs. I manually created examples for the old balance proofs that can exist for a pair of last valid balance proofs.

- added values for balance proofs edge cases, mimicking old balance proofs
- removed the tests that only tried to mimic old balance proofs with the same `transferred + claimable_locked` sum (added manual values for this); these tests are redundant and they were not implemented correctly; it is hard to implement this correctly, because there are constraints on the maximum locked amount that can exist in a valid balance proof
- added balance tests to make sure that with an old balance proof, the attacker cannot receive more tokens that he is supposed to (at the end of the channel lifecycle)

### CODE REVIEW

File: raiden_contracts/tests/test_channel_settle.py [hunk 1]
```diff
@@ -85,124 +79,72 @@ def test_settle_channel_state(
...
-        channel_test_values,
 ):
-    number_of_channels = 5
-    accounts = get_accounts(2 * number_of_channels)
-    (vals_A0, vals_B0) = channel_test_values
-
-    # We mimic old balance proofs here, with a high locked amount and lower transferred amount
-    # We expect to have the same settlement values as the original values
-
-    def equivalent_transfers(balance_proof):
```
Comment (loredanacirstea) [inline raiden_contracts/tests/test_channel_settle.py]: These mimicked balance proofs were removed, because they were not correct in the first place. There are additional constraints imposed on the locked amount, that should be `<=` total available deposit and `<=` than the available balance at a certain point in time. This is harder to mimic. Plus - they also took a lot of testing time (~10-15 min)

To replace them, I have added test values in `channel_settle_test_values` from `channel_test_values.py`. E.g.:
```
            # participant1 provides an old participant2 balance proof, but with the same
            # transferred + claimable_locked
            # This should have the same final participant balances as the valid last balance proofs
            ChannelValues(
                deposit=40,
                withdrawn=10,
                transferred=19994,
                claimable_locked=40,
                unclaimable_locked=2,
            ),
```

File: raiden_contracts/tests/test_channel_settle.py [hunk 2]
```diff
@@ -75,7 +70,6 @@ def test_settle_no_bp_success(
     ).transact({'from': A})
 
 
-@pytest.mark.parametrize('channel_test_values', channel_settle_test_values)
 def test_settle_channel_state(
```
Comment (loredanacirstea) [inline raiden_contracts/tests/test_channel_settle.py]: I left this test as a basic state test with 1 balance proof pair, following the pattern for all other tests.
Moved the more comprehensive tests for old balance proofs with various values in `test_channel_settle_unlock_state.py`, also testing the balances at the end of the channel lifecycle - so not only for `settleChannel`, but also for `unlock`.

File: raiden_contracts/tests/fixtures/channel_test_values.py [hunk 3]
```diff
@@ -1,81 +1,475 @@
...
+# We must cover the edge cases documented in
+# https://github.com/raiden-network/raiden-contracts/issues/188
+# The scope is to make sure that if someone uses an old balance proof, this cannot be used as
+# an attack to steal tokens.
+# For invalid balance proofs (created and signed with an unofficial Raiden client),
+# we cannot determine and guarantee corectness. There are specific constraints that the
+# Raiden client must enforce that guarantee correctness.
+
+# For all valid last balance proofs provided we will programatically construct a suit of balance
+# proof pairs that could be used for attacks (documented edge cases).
```
Comment (loredanacirstea) [inline raiden_contracts/tests/fixtures/channel_test_values.py]: `For all valid last balance proofs provided we will programatically construct a suit of balance` is outdated, I removed that test.

**Comment by jomuel** [comment]
Reviewing this now

File: raiden_contracts/tests/test_channel_settle_unlock_state.py [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+                    (vals_A, vals_B, 'invalid'),
+                    expected_settlement0,
+                    expected_settlement_onchain0,
+                    expected_final_balance_A0,
+                    expected_final_balance_B0,
+                )
+
+
+@pytest.mark.parametrize('channel_test_values', channel_settle_test_values)
+def test_channel_settle_old_old_balance_proof_values2(
```
Comment (jomuel) [inline raiden_contracts/tests/test_channel_settle_unlock_state.py]: As far as I can see this and the last two tests are identical except for the subsets of results they check. Can't we avoid the code duplication by turning the whole setup into a fixture, or turning the three tests into one parametrized one?

File: raiden_contracts/tests/fixtures/channel.py [hunk 5]
```diff
@@ -698,9 +704,11 @@ def get(
 
 
 def call_settle(token_network, channel_identifier, A, vals_A, B, vals_B):
-    assert vals_B.transferred + vals_B.locked >= vals_A.transferred + vals_A.locked
+    A_total_transferred = vals_A.transferred + vals_A.locked
+    B_total_transferred = vals_B.transferred + vals_B.locked
+    assert B_total_transferred >= B_total_transferred
```
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden_contracts/tests/fixtures/channel.py]: Good catch!

File: raiden_contracts/tests/test_channel_settle_unlock_state.py (hunk 4 above)
Comment (loredanacirstea) [reply raiden_contracts/tests/test_channel_settle_unlock_state.py]: About `test_channel_settle_valid_old_balance_proof_values`, `test_channel_settle_old_old_balance_proof_values` and `test_channel_settle_old_old_balance_proof_values2`:

- all the balance proof value pairs cannot be run together under the same test because travis fails with `no output for 10 min` and looking into this, I did not find a way to make travis wait except for `travis_wait`, which hides the logs until the job finishes (really not useful because you cannot pick up on failed tests early), so I just split the tested values because I thought the remaining code is not that much and did not want to add too many fixtures (add more magic to this)
- they both use `test_settlement_outcome` as a fixture for the bulk of the tests
- tried by making a fixture with the balance proof pairs, but the tests ended up running for > 2h - so we cannot do a test/test_value.

But I am open to suggestions here - I had a bit of a hard time to do the setup.

Comment (jomuel) [reply raiden_contracts/tests/test_channel_settle_unlock_state.py]: If we don't want to rewrite much or add more fixtures, what comes to my mind is a single test function with an additional argument `tested_range` that is parametrized like this:
```
@pytest.mark.parametrize('tested_range', ('valid_old', 'old_old1', 'old_old2'))

# ... setup

if tested_range is 'valid_old':
    # ... assertions from the first test
elif tested_range is 'old_old1':
    # ... assertions from the second test
else:
    # ... assertions from the third test
```
A bit unusual but also well readable... Otoh the duplicated code is really not that much and can stay as it is instead.

Comment (loredanacirstea) [reply raiden_contracts/tests/test_channel_settle_unlock_state.py]: Is it ok for now to leave it as is and you can make a PR implementing the parametrized range after?

Comment (jomuel) [reply raiden_contracts/tests/test_channel_settle_unlock_state.py]: Sure. I'll approve then.

File: raiden_contracts/tests/fixtures/channel_test_values.py [hunk 6]
```diff
@@ -1,81 +1,475 @@
...
+            # participant2 provides an old participant1 balance proof, with a higher
+            # unclaimable locked amount can happen if expired transfers are removed
+            # from the merkle tree
+            # ChannelValues(
+            #     deposit=35,
+            #     withdrawn=5,
+            #     transferred=20020,
+            #     claimable_locked=3,
+            #     unclaimable_locked=12,
+            # ),
```
[the remark this answers is not in the document]
Comment (loredanacirstea) [reply raiden_contracts/tests/fixtures/channel_test_values.py]: Opened https://github.com/raiden-network/raiden-contracts/issues/271

File: raiden_contracts/tests/fixtures/channel_test_values.py (hunk 3 above)
Comment (loredanacirstea) [reply raiden_contracts/tests/fixtures/channel_test_values.py]: Fixed in https://github.com/raiden-network/raiden-contracts/pull/272

File: raiden_contracts/tests/test_channel_settle_unlock_state.py (hunk 4 above)
Comment (loredanacirstea) [reply raiden_contracts/tests/test_channel_settle_unlock_state.py]: Opened https://github.com/raiden-network/raiden-contracts/issues/273 if you want to handle it.
